src/include/storage/metric/transaction_metric.h

```diff
@@ -0,0 +1,236 @@
+   */
+  struct TransactionData {
+    std::chrono::high_resolution_clock::time_point start_;
+    int64_t latency_;
+    int64_t tuple_read_;
+    int64_t tuple_insert_;
+    int64_t tuple_delete_;
+    int64_t tuple_update_;
+  };
+
+  /**
+   * Maps from transaction id to relevant data.
+   */
+  std::unordered_map<transaction::timestamp_t, struct TransactionData> data_;
+};
+
+/**
+ * Interface that owns and manipulates TransactionMetricRawData
+ */
+class TransactionMetric : public AbstractMetric<TransactionMetricRawData> {
+ public:
+  /**
+   * @param txn transaction context of the beginning transaction
+   */
+  void OnTransactionBegin(const transaction::TransactionContext *txn) override { GetRawData()->SetTxnStart(txn); }
+
+  /**
+   * @param txn transaction context of the committing transaction
+   * @param database_oid OID of the database the transaction is running in
+   */
+  void OnTransactionCommit(const transaction::TransactionContext *txn, catalog::db_oid_t database_oid) override {
+    (void)database_oid;
+    GetRawData()->CalculateTxnLatency(txn);
+  }
+
+  /**
+   * @param txn transaction context of the aborting transaction
+   * @param database_oid OID of the database the transaction is running in
+   */
+  void OnTransactionAbort(const transaction::TransactionContext *txn, catalog::db_oid_t database_oid) override {
+    (void)database_oid;
+    GetRawData()->CalculateTxnLatency(txn);
+  }
+
+  /**
+   * @param txn context of the transaction performing read
+   * @param database_oid OID of the database that the tuple read happens
+   * @param namespace_oid OID of the namespace that the tuple read happens
+   * @param table_oid OID of the table that the tuple read happens
+   */
+  void OnTupleRead(const transaction::TransactionContext *txn, catalog::db_oid_t database_oid,
+                   catalog::namespace_oid_t namespace_oid, catalog::table_oid_t table_oid) override {
+    (void)database_oid;
+    (void)namespace_oid;
+    (void)table_oid;
+    GetRawData()->IncrementTupleRead(txn);
+  }
+
+  /**
+   * @param txn context of the transaction performing update
+   * @param database_oid OID of the database that the tuple delete happens
+   * @param namespace_oid OID of the namespace that the tuple delete happens
+   * @param table_oid OID of the table that the tuple delete happens
+   */
+  void OnTupleUpdate(const transaction::TransactionContext *txn, catalog::db_oid_t database_oid,
+                     catalog::namespace_oid_t namespace_oid, catalog::table_oid_t table_oid) override {
+    (void)database_oid;
+    (void)namespace_oid;
+    (void)table_oid;
+    GetRawData()->IncrementTupleUpdate(txn);
+  }
+
+  /**
+   * @param txn context of the transaction performing insert
+   * @param database_oid OID of the database that the tuple delete happens
+   * @param namespace_oid OID of the namespace that the tuple delete happens
+   * @param table_oid OID of the table that the tuple delete happens
```

Comment:
Same as [previous comment](https://github.com/cmu-db/terrier/pull/338/files#r283080140). Should be `insert`

Reply:
Corrected